src/test/scala/DSATest.scala

```diff
@@ -0,0 +1,707 @@
+import analysis.{AddressRange, DSC, DSG, DSN, DataLocation, Derm, HeapLocation}
+import ir.Endian.BigEndian
+import ir.{Assign, BVADD, BinaryExpr, BitVecLiteral, CFGPosition, ConvertToSingleProcedureReturn, DirectCall, Memory, MemoryAssign, MemoryLoad, Register, SharedMemory}
+import org.scalatest.funsuite.AnyFunSuite
+import test_util.TestUtil
+import ir.dsl.*
+import specification.Specification
+import util.{BASILConfig, BoogieGeneratorConfig, ILLoadingConfig, IRContext, RunUtils, StaticAnalysisConfig}
+
+/**
+ * This is the test suite for testing DSA functionality
+ * The tests follow a general pattern of running BASIL analyses on a test program
+ * and then asserting properties about the Data Structure Graph (DSG) of the function produced at
+ * different levels
+ *
+ * DSA has three phases.
+ * BASILRESULT.analysis.get.local is the set of graphs from the end of the local phase
+ * BASILRESULT.analysis.get.bu is the set of graphs from the end of the bottom-up phase
+ * BASILRESULT.analysis.get.td is the set of graphs from the end of the top-down phase
+ *
+ */
+class DSATest extends AnyFunSuite, TestUtil {
+
+  // Local DSA tests
+  test("basic pointer") {
+    val results = RunUtils.loadAndTranslate(
+      BASILConfig(
+        loading = ILLoadingConfig(
+          inputFile = "examples/basicpointer/basicpointer.adt",
+          relfFile = "examples/basicpointer/basicpointer.relf",
+          specFile = None,
+          dumpIL = None,
+        ),
+        staticAnalysis = Some(StaticAnalysisConfig()),
+        boogieTranslation = BoogieGeneratorConfig(),
+        outputPrefix = "boogie_out",
+      )
+    )
+    val program = results.ir.program
+
+    // the dsg of the main procedure after the local phase
+    val dsg = results.analysis.get.locals.get(program.mainProcedure)
+
+
+
+    // dsg.formals(R29) is the slice representing formal R29
+    val R29formal = dsg.adjust(dsg.formals(R29))
+
+
+    // cells representing the stack at various offsets
+    val stack0 = dsg.find(dsg.stackMapping(0).cells(0)) // R31
+    val stack8 = dsg.find(dsg.stackMapping(8).cells(0)) //  R31 + 8
+    val stack40 = dsg.find(dsg.stackMapping(40).cells(0))//  R31 + 40
+    val stack32 = dsg.find(dsg.stackMapping(32).cells(0)) //  R31 + 32
+    val stack24 = dsg.find(dsg.stackMapping(24).cells(0)) //  R31 + 24 and Malloc
+
+    assert(dsg.adjust(stack0.getPointee).equals(R29formal)) // R31 points to the frame pointer
+    assert(dsg.adjust(stack8.getPointee).equals(dsg.adjust(dsg.formals(R30)))) // R31 + 8 points to the link register
+
+
+    assert(dsg.adjust(stack32.getPointee).equals(stack24))  // R31 + 32 points to R31 + 24, 00000457
```

Comment:
R31 + 32 is later set to point to a heap allocation (at 00000466), but I can't see this in the DSA results at all. How does the analysis work when a stack location is assigned multiple values during a procedure's execution?

Reply:
![image](https://github.com/user-attachments/assets/099a007a-b1cb-4d9d-9e88-73fcd5bfe345)

This is from "MACROSCOPIC DATA STRUCTURE ANALYSIS AND OPTIMIZATION" (Lattner, 2005) Figure 3.5, "local analysis function". 

A memory assignment is *Y = X is handled by merge(Ev(X), E(Ev(Y)) where "Ev: virtual register -> slice"  gives the pointee of virtual register (in this example r0 or X is replaced by tmp1 as registers are treated as SSA form and tmp2 is a implicit virtual register). E: cell -> slice gives the pointee of a cell. 
Every stack position (R31 + C) is treated as a special virtual register here.
normalised it would look like 
pos := R31 + C
*pos := tmp
merge(Ev(tmp), E(Ev(pos)), where tmp and pos  are SSA variables

now E(Ev(pos)) is either a single region/node for the whole stack such as:
Node R31 with cells at offsets: 0, 8 ... 24, 32
or treated as separate node for each offset used on the stack Node R31 + 0, Node R31 + 8, ... Node R31 + 32,
The stack access patterns are known so we can create the nodes (know their size and whether two nodes have overlapped accesses in order to merge them to one region/node prior). 

if a stack address is stored in another position in memory
mem = 0x000001
*mem = R31 + 8

we will have to to merge(Ev(R31 + 8), E(Ev(mem))), if the stack is treated as a single node, the node pointed to by the memory location mem is unified with the node of the stack. This will cause the stack to lose all field sensitivity eventually since the cell E(Ev(mem) is unlikely to have node that structurally is equivalent to the stack.
However, if multiple regions are created this is avoided. This is still sound if we have correctly found all the stack accesses and merged any overlapping R31 + C accesses into one. 

Therefore tmp := R31 + C is equivalent to X = alloca in the figure above.  that is we have 
Ev(tmp) = new node(R31 + C).0 where node(R31 + C).0 is the cell at offset 0 of the new node which is marked as a stack region. Any reads/writes to the stack position R31 + C, now is represented by E(Ev(tmp))

Position 0000044F: tmp1 := R31 + 24 // Ev(tmp1) = new Node(R31 + 24).0
// implicit normalisation: tmp2 := R31 + 32  // Ev(tmp2) = new Node(R31 + 32).0
Position 00000457:  *tmp2  := tmp1 // merge(Ev(tmp1), E(Ev(tmp2)))
// Therefore, Node(R31 + 32).0.pointee is merged with  Node(R31 + 24).0, making E(Node(R31 + 32).0) == Node(R31 + 24).0

at position 00000446 Ev(R0) == Malloc_Node  == E(Node(R31 + 24).0)
we have 
mem := mem with [R31 + 0x20, el]:u64 <- R0
*(R31 + 32) := R0
merge(Ev(R0), E(Ev(R31+ 32)) 
== merge(E(Node(R31 + 24).0),  E(Node(R31 + 32).0)) 
== merge(E(Node(R31 + 24).0), Node(R31 + 24).0)
which merges make the stack + 24 point to itself

In general any memory location which has multiple values assigned to it, is has its pointee merged with all of them.
The analysis is flow insensitive, therefore, the order of assignment don't matter

Reply:
I guess what I didn't understand is why only some stack stores seem to appear in nodes. Is it just that the first stack store at each offset has a node created for it - if so, why refer to a particular MemoryStore in the node?

Reply:
This is allows us to know which CFGPosittion(s) created the current node and was mainly used for debugging